p2p/transport/webrtc/listener.go

```diff
@@ -0,0 +1,281 @@
+		localFingerprint:          localFingerprints[0],
+		localFingerprintMultibase: localFpMultibase,
+		localMultiaddr:            laddr,
+		ctx:                       ctx,
+		cancel:                    cancel,
+		localAddr:                 socket.LocalAddr(),
+		connChan:                  make(chan tpt.CapableConn, maxConnectionBufferSize),
+	}
+
+	l.wg.Add(1)
+	go l.handleIncomingCandidates(candidateChan)
+	return l, err
+}
+
+func (l *listener) handleIncomingCandidates(candidateChan chan candidateAddr) {
+	defer l.wg.Done()
+	for {
+		select {
+		case <-l.ctx.Done():
+			return
+		case addr := <-candidateChan:
+			go func() {
+				ctx, cancelFunc := context.WithTimeout(context.Background(), candidateSetupTimeout)
+				defer cancelFunc()
+				conn, err := l.handleCandidate(ctx, addr)
+				if err != nil {
+					log.Debugf("could not accept connection: %v", err)
+					return
+				}
+				l.connChan <- conn
+			}()
+		}
+	}
+}
+
+func (l *listener) Accept() (tpt.CapableConn, error) {
+	select {
+	case <-l.ctx.Done():
+		return nil, os.ErrClosed
+	case conn := <-l.connChan:
+		return conn, nil
+	}
+}
+
+func (l *listener) Close() error {
+	select {
+	case <-l.ctx.Done():
+		return nil
+	default:
+	}
+	l.cancel()
+	l.wg.Wait()
+	return nil
+}
+
+func (l *listener) Addr() net.Addr {
+	return l.localAddr
+}
+
+func (l *listener) Multiaddr() ma.Multiaddr {
+	return l.localMultiaddr
+}
+
+func (l *listener) handleCandidate(ctx context.Context, addr candidateAddr) (tpt.CapableConn, error) {
+	remoteMultiaddr, err := manet.FromNetAddr(addr.raddr)
+	if err != nil {
+		return nil, err
+	}
+	scope, err := l.transport.rcmgr.OpenConnection(network.DirInbound, false, remoteMultiaddr)
+	if err != nil {
+		return nil, err
+	}
+	pc, conn, err := l.setupConnection(ctx, scope, remoteMultiaddr, addr)
+	if err != nil {
+		scope.Done()
+		if pc != nil {
+			_ = pc.Close()
+		}
+		return nil, err
+	}
+	return conn, nil
+}
+
+func (l *listener) setupConnection(ctx context.Context, scope network.ConnManagementScope, remoteMultiaddr ma.Multiaddr, addr candidateAddr) (*webrtc.PeerConnection, tpt.CapableConn, error) {
+
+	settingEngine := webrtc.SettingEngine{}
+	settingEngine.SetAnsweringDTLSRole(webrtc.DTLSRoleServer)
+	settingEngine.SetICECredentials(addr.ufrag, addr.ufrag)
+	settingEngine.SetLite(true)
+	settingEngine.SetICEUDPMux(l.mux)
+	settingEngine.DisableCertificateFingerprintVerification(true)
+	settingEngine.DetachDataChannels()
+
+	api := webrtc.NewAPI(webrtc.WithSettingEngine(settingEngine))
+
+	pc, err := api.NewPeerConnection(l.config)
+	if err != nil {
+		return pc, nil, err
+	}
+
+	signalChan := make(chan error)
+	var wrappedChannel *dataChannel
+	var handshakeOnce sync.Once
+
+	handshakeChannel, err := pc.CreateDataChannel("", &webrtc.DataChannelInit{
+		Negotiated: func(v bool) *bool { return &v }(true),
+		ID:         func(v uint16) *uint16 { return &v }(0),
+	})
+	if err != nil {
+		return pc, nil, err
+	}
+
+	handshakeChannel.OnOpen(func() {
```

Comment:
I'm not really sure when exactly this is called (a comment would help), but it's probably easier to _just_ pass a signal into a channel and do the rest of the setup outside of the callback. That way, you don't have to pass `wrappedChannel` AND the error out.

Reply:
We can't call Detach outside OnOpen. https://github.com/pion/webrtc/blob/master/datachannel.go#L192-L216. I need the detached channel for stream level congestion control. Here, onOpen fires immediately when the handler is set since this is a negotiated datachannel i.e. it it already in the open state. https://github.com/pion/webrtc/blob/7ab3174640b3ce15abebc2516a2ca3939b5f105f/datachannel.go#L203